drv/fpga-devices/src/ecp5.rs

```diff
@@ -0,0 +1,453 @@
+impl<ImplT: Ecp5Impl> Ecp5<ImplT> {
+    /// Send a command to the device which does not return or require additional
+    /// data. FPGA-TN-02039-2.0, 6.2.5 refers to this as a Class C command.
+    pub fn send_command(&self, c: Command) -> Result<(), ImplT::Error> {
+        self.0.lock()?;
+        self.0.write_command(c)?;
+        self.0.release()?;
+        ringbuf_entry!(Trace::Command(c));
+        Ok(())
+    }
+
+    /// Send a command and read back a number of bytes given by the type T. Note
+    /// that data is always returned in big endian order.
+    pub fn read<T: Default + AsBytes + FromBytes>(
+        &self,
+        c: Command,
+    ) -> Result<T, ImplT::Error> {
+        let mut buf = T::default();
+
+        self.0.lock()?;
+        self.0.write_command(c)?;
+        self.0.read(buf.as_bytes_mut())?;
+        self.0.release()?;
+
+        Ok(buf)
+    }
+
+    /// Send a `Command` and read back two bytes of data.
+    pub fn read16(&self, c: Command) -> Result<u16, ImplT::Error> {
+        Ok(u16::from_be(self.read(c)?))
+    }
+
+    /// Send a `Command` and read back four bytes of data.
+    pub fn read32(&self, c: Command) -> Result<u32, ImplT::Error> {
+        let v = u32::from_be(self.read(c)?);
+
+        match c {
+            Command::ReadId => {
+                ringbuf_entry!(Trace::ReadId(v, Id::from(v)))
+            }
+            Command::ReadStatus => ringbuf_entry!(Trace::ReadStatus(v)),
+            Command::ReadUserCode => ringbuf_entry!(Trace::ReadUserCode(v)),
+            _ => ringbuf_entry!(Trace::Read32(c, v)),
+        }
+
+        Ok(v)
+    }
+
+    /// Read the Status register
+    pub fn status(&self) -> Result<Status, ImplT::Error> {
+        Ok(Status(self.read32(Command::ReadStatus)?))
+    }
+
+    /// Enable ConfigurationMode, allowing access to certain configuration
+    /// command and the bitstream loading process.
+    pub fn enable_configuration_mode(&self) -> Result<(), ImplT::Error> {
+        self.send_command(Command::EnableConfigurationMode)
+    }
+
+    /// Leave ConfigurationMode, disabling access to certaion commands. In
+    /// addition, if a bitstream was loaded, this will transition the device to
+    /// UserMode.
+    pub fn disable_configuration_mode(&self) -> Result<(), ImplT::Error> {
+        self.send_command(Command::DisableConfigurationMode)
+    }
+
+    /// Wait for the device to finish an operation in progress, sleeping for the
+    /// given duration between polling events.
+    pub fn await_not_busy(&self, sleep_ticks: u64) -> Result<(), ImplT::Error> {
+        while self.status()?.busy() {
+            hl::sleep_for(sleep_ticks);
+        }
+        Ok(())
+    }
+
+    /// Wait for the DONE flag to go high.
+    pub fn await_done(&self, sleep_ticks: u64) -> Result<(), ImplT::Error> {
+        while !self.status()?.done() {
+            hl::sleep_for(sleep_ticks);
+        }
+        Ok(())
+    }
+}
+
+impl<ImplT: Ecp5Impl> From<ImplT> for Ecp5<ImplT> {
```

Comment:
This seems weird to me – are there places where we want to use `from` / `into` instead of the plain `Ecp5(...)` constructor?